src/main/java/org/zalando/nakadi/service/ConsumerLimitingService.java

```diff
@@ -0,0 +1,182 @@
+import org.zalando.nakadi.exceptions.NakadiRuntimeException;
+import org.zalando.nakadi.exceptions.NoConnectionSlotsException;
+import org.zalando.nakadi.exceptions.ServiceUnavailableException;
+import org.zalando.nakadi.repository.zookeeper.ZooKeeperHolder;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.Random;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import static java.text.MessageFormat.format;
+import static org.echocat.jomon.runtime.concurrent.Retryer.executeWithRetry;
+
+@Service
+public class ConsumerLimitingService {
+
+    public static final String CONNECTIONS_ZK_PATH = "/nakadi/consumers/connections";
+
+    private static final Logger LOG = LoggerFactory.getLogger(ConsumerLimitingService.class);
+
+    private static final String ERROR_MSG = "You exceeded the maximum number of simultaneous connections to a single " +
+            "partition for event type '{0}', partition(s): {1}; max limit is {2} connections per client";
+
+    private final ZooKeeperHolder zkHolder;
+    private final int maxConnections;
+    private final List<String> slotNames;
+
+    @Autowired
+    public ConsumerLimitingService(final ZooKeeperHolder zkHolder,
+                                   @Value("${nakadi.stream.maxConnections}") final int maxConnections) {
+        this.zkHolder = zkHolder;
+        this.maxConnections = maxConnections;
+
+        slotNames = IntStream.range(0, maxConnections)
+                .boxed()
+                .map(String::valueOf)
+                .collect(Collectors.toList());
+    }
+
+    @SuppressWarnings("unchecked")
+    public List<ConnectionSlot> acquireConnectionSlots(final String client, final String eventType,
+                                                       final List<String> partitions)
+            throws NoConnectionSlotsException, ServiceUnavailableException {
+
+        final List<Optional<ConnectionSlot>> slots = new ArrayList<>();
+        try {
+            // acquire slots for all partitions
+            for (final String partition : partitions) {
+                Optional<ConnectionSlot> connectionSlot = Optional.empty();
+                try {
+                    connectionSlot = executeWithRetry(
+                            () -> acquireConnectionSlot(client, eventType, partition),
+                            new RetryForSpecifiedCountStrategy<Optional<ConnectionSlot>>(maxConnections)
+                                    .withExceptionsThatForceRetry(ConnectionSlotOccupiedException.class)
+                                    .withWaitBetweenEachTry(0L, 100L));
+                } catch (ConnectionSlotOccupiedException e) {
+                    LOG.info(format("Failed to capture consuming connection slot after 5 tries for client '{0}', " +
+                            "event-type '{1}', partition {2}", client, eventType, partition));
+                }
+                slots.add(connectionSlot);
```

Comment:
I do that because I want to have a full list of partitions that don't have free slots - to give user the full information.